Implement transport socket match.

incfly opened the pull request:
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description:
Risk Level: Medium (Opt-in required)
Testing: unit test, integration tested.
Docs Changes: TODO(incfly) add an architecture docs.
Release Notes: New feature, implement `Cluster.transport_socket_matches`. Envoy can be configured to use different transport socket configuration to different endpoints based on metadata match. 
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/8016
[Optional Deprecated:]

incfly commented:
Still WIP PR, but will be great to get early feedback, so open a draft PR, @htuch @lizan @PiotrSikora thanks! 

I'll continue to work on

- Integration test, thus the demo `incfly/xxx.yaml` will be removed before the merge.
- Docs in the architecture about detailed use case.

htuch commented on incfly/eds.yaml:
```diff
@@ -0,0 +1,17 @@
+version_info: "0"
```
I assume these are not going to be checked in..

incfly replied:
removed. superseded by the integration test.

htuch commented on source/common/upstream/transport_socket_matcher.cc:
```diff
@@ -0,0 +1,85 @@
+  }
+}
+
+bool metadataMatch(const envoy::api::v2::core::Metadata& metadata,
```
Why can't we reuse the matching code that already exists for matching and is part of the subset LB?

incfly replied:
I think this [`hostMatches`](https://github.com/envoyproxy/envoy/blob/f51fab0d4518e87ae93d71b3c42f8c69d68a9514/source/common/upstream/subset_lb.cc#L437) is very closed to the `metadataMatch` logic here.

I'm thinking maybe we can extract that part of code into a shared library/utility function, used by both `transport_socket_match` and `subset_lb`?

Or you suggest we extend the subset_lb to be able to select socket implementation at all? If so, I feel the `subset_lb` name is not a good fit and mixed with load balancing with transport socket logic?

htuch replied:
Yeah, that's what I was thinking, some kind of utility to do metadata matching that is shared by both subset LB and the transport socket matcher, similar to `hostMatches()`.

incfly replied:
done. moved to config/metadata/metadata.cc

incfly commented on include/envoy/upstream/upstream.h:
```diff
@@ -781,6 +781,14 @@ class ClusterInfo {
   virtual Network::TransportSocketFactory& transportSocketFactory() const PURE;
 
+  /**
+   * @return Network::TransportSocketFactory& the factory of transport socket to use when
```
done.

incfly commented:
Added integration test and removed the samples files, moved from draft to normal PR.

lizan commented on source/common/upstream/BUILD:
```diff
@@ -373,6 +373,40 @@ envoy_cc_library(
+    name = "transport_socket_matcher_lib",
+    srcs = ["transport_socket_matcher.cc"],
+    deps = [
+        ":eds_lib",
```
do you need all of them? Can you strip these down to what you need?

incfly replied:
done. removed unnecessary ones.

incfly commented on include/envoy/upstream/upstream.h:
```diff
@@ -784,6 +784,17 @@ class ClusterInfo {
   virtual Network::TransportSocketFactory& transportSocketFactory() const PURE;
 
+  /**
+   * @param address is the upstream networking address trying to connect to.
```
because we just need address as debugging log, not affecting the behavior.

incfly commented:
how we should name this? I feel `envoy.transport_socket_match` is better than `envoy.transport_socket`, which is the prefix of socket types, `envoy.transport_sockets.alts`.

another thing, I'm changing this to default behavior, and relying on the empty `transport_socket_matches` return the default factory, to retain the behavior. let me know if this is too risky or not.

incfly commented:
All required tests passed now, PTAL, thanks!

lizan commented on include/envoy/upstream/upstream.h:
```diff
@@ -778,11 +778,29 @@ class ClusterInfo {
+   * @return Network::ClusterInfo::TransportSocketFactoryOption the default option
+   *         to select a transport socket for a given cluster.
+   */
+  static TransportSocketFactoryOption DefaultTransportSocketFactoryOption() {
```
nit: `defaultTransportSocketFactoryOption`

incfly replied:
will do.

incfly replied:
no longer needed since using `absl::optional`.

lizan commented on include/envoy/upstream/upstream.h:
```diff
@@ -778,11 +778,29 @@ class ClusterInfo {
    */
-  virtual Network::TransportSocketFactory& transportSocketFactory() const PURE;
+  virtual Network::TransportSocketFactory&
+  transportSocketFactory(const TransportSocketFactoryOption& option =
```
no default argument in virtual method. https://google.github.io/styleguide/cppguide.html#Default_Arguments

incfly replied:
Sure. but then we have two options

1. update all `transportSocketFactory()` callers to invoke this `defaulTransportSocketFactoryOption` as argumetn
1. create an overloading function with 0 argument, basically two `transportSocketFactory`, so existing invocations won't have to be modified.

which you think is better? second one obvious results in smaller changes though.

incfly replied:
hmm, speaking of that, I just realized we have two more call sites of the `ClusterInfo::transportSocketFactory`. 

1. [Router Filter](https://github.com/envoyproxy/envoy/blob/eff020170c6267e6c8dc235473f7fc85c5b1e07d/source/common/router/router.cc#L94)
2. [Dynamic Forward proxy filter](https://github.com/envoyproxy/envoy/blob/a659c5ecfd93a9d3a6580cbc968aa3fcca8eb480/source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc#L59)

Their purpose is only checking socket `implementsSecureTransport` though, but I  think to ensure features working all the cases, somehow we have to return the same type of socket factory as upstream_impl choose?

lizan replied:
either way works for default arguments, mildly prefer 1 with changing the argument to absl::optional so the default will be absl::nullopt.

for implementsSecureTransport I guess this should reflect the real upstream transport socket selection. Don't worrying about Dynamic Forward proxy filter, because that one requires dynamic forward cluster, so EDS is not in consideration. For Router Filter, I guess we already know which transport socket factory will be used when transportSocketFactory is called?

incfly replied:
`Router::FilterUtility::setUpstreamScheme` as a static function is invoked in router filter.

My understanding of the router's role

1. downstream request processed by filter
1. router process request and route based on path/xxx to different cluters.
1. (maybe) load balancing decision is made after route? what's the name in envoy component?
1. http connection pool is being selected.
1. connection is being established (`upstream_impl.cc`)

So by the time router is requesting the resolution of the `transportSocketFactory`, it does not know which upstream endpoint is connected to... 

However, I looked at the usage, it merely look at `transportSocketFactory().implementsSecureTransport()`. based on this decision, post processed the url (remove port 80 or 443...)

So seems like this won't affect connection with upstream? And to be consistent, can we just add an `ClusterInfo::implementSecureTransport`, such that it returns true if all the socket types return `implementSecureTransport` true?

This is consistent with our [discussion in the doc](https://docs.google.com/document/d/1lpshO7E-2KQ5SC519oiRe25TPAXp1HUcyVzZ4QM4IeU/edit?disco=AAAADb9FJUk)

lizan replied:
It does affect, I believe upstream HTTP implementations will reject the request with incorrect scheme (i.e. sending :scheme: http to HTTPS endpoint or vice versa).

It seems like in router.cc, Router::FilterUtility::setUpstreamScheme is called at https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L505, which after ConnPool selecting at https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L453, so you should be able to know which host is selected there. It might need some sort of refactoring however.

incfly replied:
okay, now it's addressed.

1. router part conn pool is selected by connection is created yet, so has access to host metadata, which is used to resolve the transport socket factory as the same upstream is doing.
2. `health_checker_impl` is different, it creates connection, so having access to the connection/codecclient, but not the factory/host information, thus I end up adding `implementSecureTransports` in transport socket class.

lizan commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -39,6 +39,8 @@
 
 #include "extensions/transport_sockets/well_known_names.h"
 
+#include "spdlog/spdlog.h"
```
no spdlog outside of logger library, I don't think this is directly used?

incfly replied:
right, leftover of debugging..

incfly replied:
resolving.

incfly commented:
All tests passed now with comments addressed, this is now ready. PTAL, thanks.

lizan commented on include/envoy/network/connection.h:
```diff
@@ -313,6 +313,11 @@ class ClientConnection : public virtual Connection {
+  /**
+   * Returns true if the underlying transport is secure.
+   */
+  virtual bool isSecureTransport() const PURE;
```
is this only for health checker? If so I think it might make more sense to let `createHealthCheckConnection` return whether the connection is secured?

incfly replied:
Health checker uses the codec_client as member, and bind a connection instance with that.

`createHealthCheckConnection` is only invoked during [initialization](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/health_checker_impl.cc#L200, but later on every request we need to do this `setupUpstreamScheme` thing, with only access to [`codec_client_`](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/health_checker_impl.cc#L208)

lizan replied:
I know, but you can cache the value from connection creation there. That won't change after connection initialization.

incfly replied:
> cache the value from connection there,

do you mean cache into `health_check_impl` itself? or codec_client?

Trying to understand two comments together, is the end goal to not adding `implementsSecureTransport` to both `TransportSocketFactory`? Which is kinda redundant?

lizan replied:
in short term (while health checker is not using connpool, there's a TODO there) have a flag in `health_check_impl` should be enough.

incfly replied:
not needed since now we make `transportSocketFactory` part of `HostDesription`, we'll just invoke that directly.

lizan commented on source/common/router/router.cc:
```diff
@@ -502,7 +501,16 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   route_entry_->finalizeRequestHeaders(headers, callbacks_->streamInfo(),
                                        !config_.suppress_envoy_headers_);
-  FilterUtility::setUpstreamScheme(headers, *cluster_);
+  envoy::api::v2::core::Metadata host_metadata;
```
rather than resolving transportSocketFactory again, can this get from the client connection from ConnPool? (having a flag in conn pool seems fine)

incfly replied:
The transport socket factory is actually not resolved during the pool selection call chain. Right now it only occurs as part of the connection event, when upstream is about to make connection. At this point, no client connection is being created.

1 pool is corresponding to 1 host (selected), I could add a flag to conn pool. This would require to plumb/add `ClusterInfo` reference in [ClusterManagerImpl](https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/source/common/upstream/cluster_manager_impl.cc#L1269)  to the conn pool constructor though, since resolving socket factory requires

- `ClusterInfo::transport_socket_match_`
- host::metadata (which the pool has as part of the constructor already.
- And then we set a constant member field into `HttpConnectionPool`, and able to return with a accessor method.

Does above sound good to you?

lizan replied:
Hmm, I thought the host selection resolves the transport socket factory, no?

incfly replied:
Actually, I just realized the Host itself already has access to ClusterInfo, plus it owns metadata and address as well, so `HostDescription` itself can just provide this information.

But extra invocation of the `TransportSocketMatcher::resolve` could mess up with the stats though. maybe add extra param like `bool create_connection`, and only increment stats when `create_connection` is true?

lizan replied:
Yeah HostDescription seems a perfect place, as well as to do `TransportSocketMatcher::resolve`. I don't think you should mess stats and the `bool create_connection` flag, currently the transport socket resolving is pretty static so don't even have to be per-connection basis, it can be only done once per `Host/HostDescription`, right?

incfly replied:
that's right.

> transport socket resolving is pretty static 

sounds good. I can add `TransporSocketMatcher::statsByHost()` and then HostDescription constructor will be responsible for incrementing that stats entry.

The counter will be per host basis.

lizan replied:
sgtm

incfly replied:
I end up adding `transportSocketFactory()` for the `HostDesription`. I just realize we shouldn't go through match loop iteration every time we make a connection. Instead we can just do it once, avoiding the computation.

I think this is okay, even the CDS push updates occur, Envoy should recycle those old Host object, right?

lizan replied:
> I think this is okay, even the CDS push updates occur, Envoy should recycle those old Host object, right?

If cluster didn't change yes for sure, not sure in the case if cluster changed, though you'll need to recalculate the match so it doesn't matter either.

lizan commented on include/envoy/network/transport_socket.h:
```diff
@@ -91,6 +91,11 @@ class TransportSocket {
+  /**
+   * @return bool whether the transport socket implements secure transport.
+   */
+  virtual bool implementsSecureTransport() const PURE;
```
I don't think this one is used either, the one in TSF is, but this one isn't.

incfly commented on include/envoy/upstream/host_description.h:
```diff
@@ -100,6 +101,8 @@ class HostDescription {
    */
   virtual const std::string& hostname() const PURE;
 
+  virtual Network::TransportSocketFactory& transportSocketFactory() const PURE;
```
done.

incfly commented on test/mocks/network/transport_socket.h:
```diff
@@ -0,0 +1,45 @@
+#include <string>
+#include <vector>
+
+// #include "connection.h"
```
done.

incfly commented on source/common/upstream/transport_socket_match_impl.h:
```diff
@@ -0,0 +1,59 @@
+    std::string name;
+    Network::TransportSocketFactoryPtr factory;
+    Config::Metadata::LabelSet label_set;
+    mutable TransportSocketMatchStats stats; // TODO: why?
```
this is a todo question to myself i found without mutable i have trouble passing stats to construct the struct.

removed.

incfly commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -244,13 +245,16 @@ void HostImpl::setEdsHealthFlag(envoy::api::v2::core::HealthStatus health_status
 HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& cluster,
-                           Network::Address::InstanceConstSharedPtr address,
+                           const Network::Address::InstanceConstSharedPtr& address,
+                           //  const envoy::api::v2::core::Metadata& metadata,
```
done.

incfly commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -587,8 +589,8 @@ class FactoryContextImpl : public Server::Configuration::CommonFactoryContext {
     const envoy::api::v2::Cluster& config, const envoy::api::v2::core::BindConfig& bind_config,
-    Runtime::Loader& runtime, Network::TransportSocketFactoryPtr&& socket_factory,
-    Stats::ScopePtr&& stats_scope, bool added_via_api,
+    Runtime::Loader& runtime, /*Network::TransportSocketFactoryPtr&& socket_factory,*/
```
done.

incfly commented on source/common/upstream/upstream_impl.h:
```diff
@@ -512,9 +528,9 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
-                  Network::TransportSocketFactoryPtr&& socket_factory,
-                  Stats::ScopePtr&& stats_scope, bool added_via_api,
-                  ProtobufMessage::ValidationVisitor& validation_visitor,
+                  /*Network::TransportSocketFactoryPtr&& socket_factory,*/
```
done

incfly commented on test/common/upstream/BUILD:
```diff
@@ -375,6 +375,33 @@ envoy_cc_test(
+        "//include/envoy/upstream:cluster_manager_interface",
+        "//source/common/config:metadata_lib",
+        "//source/common/event:dispatcher_lib",
+        "//source/common/json:config_schemas_lib",
```
done.

incfly commented:
Addressed the commands and fixed some known broken test.

Sorry I used the force push because seems like the commit of auto apply suggestions do not have correct DCO.

lizan commented on test/mocks/upstream/transport_socket_match.cc:
```diff
@@ -0,0 +1,26 @@
+
+namespace Envoy {
+namespace Upstream {
+namespace Outlier {
```
erh, why this in Outlier??

incfly replied:
oh, i was confused as well before. copied from other part, now I realized that only enclose portion of the header files.

Removed the outlier.

lizan commented on source/common/upstream/transport_socket_match_impl.h:
```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <array>
```
strip includes

incfly replied:
done.

incfly commented:
/retest

incfly commented:
/azp run envoy-linux

incfly commented:
All required test passed now, ready for another look.

lizan commented on source/common/upstream/upstream_impl.h:
```diff
@@ -89,7 +92,20 @@ class HostDescriptionImpl : virtual public HostDescription {
   }
 
-  // Upstream::HostDescription
+  Network::TransportSocketFactory&
```
this two method (constructor and resolveTransportSocketFactory) are long enough, can you move it to .cc?

incfly replied:
done. moved.

lizan commented on source/common/upstream/upstream_impl.h:
```diff
@@ -186,8 +203,9 @@ class HostImpl : public HostDescriptionImpl,
   CreateConnectionData createConnection(
       Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
-      Network::TransportSocketOptionsSharedPtr transport_socket_options) const override;
+      const Network::TransportSocketOptionsSharedPtr transport_socket_options) const override;
```
top level const parameter doesn't make any change, make it const ref or drop const.

incfly replied:
const dropped.

lizan commented on source/common/upstream/upstream_impl.h:
```diff
@@ -75,7 +77,8 @@ class HostDescriptionImpl : virtual public HostDescription {
         locality_zone_stat_name_(locality.zone(), cluster->statsScope().symbolTable()),
-        priority_(priority) {
+        priority_(priority), socket_factory_(resolveTransportSocketFactory(
+                                 dest_address ? dest_address->asString() : "empty", metadata)) {
```
I will move the logic ` dest_address ? dest_address->asString() : "empty"` into `resolveTransportSocketFactory`, so it is not evaluated other than debug logging is enabled.

incfly replied:
done. interesting to see the power of the ENVOY_LOG macro.

lizan commented on source/common/upstream/upstream_impl.h:
```diff
@@ -60,36 +61,25 @@ class HealthCheckHostMonitorNullImpl : public HealthCheckHostMonitor {
-            : Network::Utility::getAddressWithPort(*dest_address, health_check_config.port_value());
+      uint32_t priority);
+
+  Network::TransportSocketFactory&
```
move this to private? seems not used outside

incfly replied:
done.

incfly commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -217,10 +217,46 @@ HostVector filterHosts(const std::unordered_set<HostSharedPtr>& hosts,
+}
+
+Network::TransportSocketFactory& HostDescriptionImpl::resolveTransportSocketFactory(
+    Network::Address::InstanceConstSharedPtr& dest_address,
```
done.

incfly commented:
/retest

lizan commented on source/common/upstream/upstream_impl.h:
```diff
@@ -606,7 +597,8 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   absl::optional<std::chrono::milliseconds> idle_timeout_;
   const uint32_t per_connection_buffer_limit_bytes_;
-  Network::TransportSocketFactoryPtr transport_socket_factory_;
+  // Network::TransportSocketFactoryPtr transport_socket_factory_;
```
delete

incfly replied:
done.

incfly commented:
Many thanks @lizan !
